Fix #7681: HTML Select Option Styling

U8NWXD opened the pull request:
## Explanation

Fixes #7681

Boostrap 4 uses code like this for dropdown menus:

```html
<div class="dropdown-menu" aria-labelledby="dropdownMenuButton">
    <a class="dropdown-item" href="#">Action</a>
    <a class="dropdown-item" href="#">Another action</a>
    <a class="dropdown-item" href="#">Something else here</a>
</div>
```

With the upgrade from Bootstrap 3.3.4 to 4.3.1 in ba55ea9, the syntax
for dropdown menus changed to include the `dropdown-item` class for menu
items. The `html-select.directive.html` file was not updated, so this
commit adds the `dropdown-item` class to it.

Buggy UI:

![Screen Shot 2019-09-22 at 18 48 07](https://user-images.githubusercontent.com/19878639/65398226-dae19700-dd6a-11e9-932a-9a182cb64023.png)

Old (Release 2.8.5) UI:

![](https://user-images.githubusercontent.com/19878639/65976864-b306c900-e425-11e9-8487-73bf3856a70c.png)

New UI:

![Screen Shot 2019-10-01 at 08 28 31](https://user-images.githubusercontent.com/19878639/66413238-dc869e00-e9ab-11e9-831d-c35503e8f43f.png)

Fixes #7681.

## Checklist
- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The linter/Karma presubmit checks have passed.
  - These should run automatically, but if not, you can manually trigger them locally using `python -m scripts.pre_commit_linter` and `python -m scripts.run_frontend_tests`.
- [x] The PR is made from a branch that's **not** called "develop".
- [ ] The PR has an appropriate "PROJECT: ..." label (Please add this label for the first-pass review of the PR).
- [x] The PR has an appropriate "CHANGELOG: ..." label (If you are unsure of which label to add, ask the reviewers for guidance).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide).
- [x] The PR addresses the points mentioned in the codeowner checks for the files/folders changed. (See the [codeowner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).)
- [ ] The PR is **assigned** to an appropriate reviewer.
  - If you're a new contributor, please ask on [Gitter](https://gitter.im/oppia/oppia-chat) for someone to assign a reviewer and don't tick this checkbox.
  - If you're not sure who the appropriate reviewer is, please assign to the issue's "owner" -- see the "talk-to" label on the issue. Do not only request the review but also add the reviewer as an assignee.

U8NWXD committed:
Fix HTML Select Option Styling

Add CSS styling for the HTML select option dropdowns, such as that which
appears when setting the answer for a multiple-choice interaction
response. This styling restores the spacing and
background-color-on-hover currently in production.

Fixes #7681.

U8NWXD committed:
Move HTML Select Option Styling to Directive HTML

Move the CSS styling for HTML select dropdown options to the directive
HTML file as required by the
[style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide#css).

Fixes #7681.

DubeySandeep commented on core/templates/dev/head/components/forms/custom-forms-directives/html-select.directive.html:
```diff
@@ -16,6 +16,13 @@
+    display: block;
+    padding: 3px 20px;
+  }
+  .oppia-html-select > ul > li > a:hover {
```
Is it similar to the CSS on line 58? (Do you know why this issue happened in the first place, although it was working fine earlier?)

U8NWXD replied:
It looks like the version live on https://oppia.org right now gets the styling I added from `third_party.css`, so perhaps `third_party.css` changed. In this screenshot you can see that the sizing and background color styling comes from `third_party.min.[...].css`:

![Screen Shot 2019-10-01 at 08 34 01](https://user-images.githubusercontent.com/19878639/65977972-b438f580-e427-11e9-9805-0f974d69fb70.png)

I am not sure what the CSS on line 58 is doing. It applies to the text of the answer choice, not the box around it. It also seems to apply a blue background on hover, which I do not see occurring. Here you can see the line 58 styling applied to the text of the answer choice:

![Screen Shot 2019-10-01 at 08 53 15](https://user-images.githubusercontent.com/19878639/65978836-3d9cf780-e429-11e9-9435-de5c4d434bfe.png)

U8NWXD replied:
@vojtechjelinek are you familiar with how `third_party.css` is generated? I'm not sure where to check for changes that might have introduced this bug since I think `third_party.css` is generated by the `scripts.start` script (so no `git blame`)

DubeySandeep replied:
@U8NWXD, I think this issue might have introduced because of some upgrade in third_party library,  third_party.css file is a minified version of all the third_party style sheets. I think you can find the comment which has information about the third_party which is introducing the given style to the class/id. (similar to the image below)

![image](https://user-images.githubusercontent.com/16653571/66154618-888e5900-e63b-11e9-941f-e5678417740e.png)

This info will help you find the cause of the issue (maybe).

U8NWXD replied:
This bug appears to have been introduced by the migration to Bootstrap 4.3.1 from 3.3.4 (ba55ea9b5df8d6c8d3d2b82ca0c11800f8409d81), which was not included in the [2.8.5 release](https://github.com/oppia/oppia/commits/release-2.8.5). Bootstrap 4 now uses syntax like this for dropdowns:

```html
<div class="dropdown">
  <button class="btn btn-secondary dropdown-toggle" type="button" id="dropdownMenuButton" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false">
    Dropdown button
  </button>
  <div class="dropdown-menu" aria-labelledby="dropdownMenuButton">
    <a class="dropdown-item" href="#">Action</a>
    <a class="dropdown-item" href="#">Another action</a>
    <a class="dropdown-item" href="#">Something else here</a>
  </div>
</div>
```
(see [bootstrap docs](https://getbootstrap.com/docs/4.3/components/dropdowns/)).

We use Angular UI Bootstrap, which is [only compatible with Bootstrap 3.x](https://angular-ui.github.io/bootstrap/). The dropdown menu is created by `core/templates/dev/head/components/forms/custom-forms-directives/html-select.directive.html`, which includes this code:

```html
  <ul uib-dropdown-menu style="max-width: 80%; overflow: auto; width: auto;">
    <li ng-repeat="choice in options" style="float: left;">
      <a ng-click="select(choice.id)">
        <angular-html-bind class="oppia-html-select-option protractor-test-html-select-option" html-data="choice.val">
        </angular-html-bind>
      </a>
    </li>
  </ul>
```

`uib-dropdown-menu` is an Angular UI Bootstrap directive that generates code like this:

```html
  <ul uib-dropdown-menu="" style="max-width: 80%; overflow: auto; width: auto; position: absolute; will-change: transform; top: 0px; left: 0px; transform: translate3d(0px, 30px, 0px);" class="dropdown-menu show" x-placement="bottom-start">
    <li ng-repeat="choice in options" style="float: left;" class="">
      <a ng-click="select(choice.id)">
        <angular-html-bind class="oppia-html-select-option protractor-test-html-select-option" html-data="choice.val"><p>pass</p></angular-html-bind>
      </a>
    </li>
    <li ng-repeat="choice in options" style="float: left;" class="">
      <a ng-click="select(choice.id)">
        <angular-html-bind class="oppia-html-select-option protractor-test-html-select-option" html-data="choice.val"><p>fail</p></angular-html-bind>
      </a>
    </li>
  </ul>
```

If we add `dropdown-item` as a class for the `<li>` elements like this:

```html
  <ul uib-dropdown-menu style="max-width: 80%; overflow: auto; width: auto;">
    <li ng-repeat="choice in options" style="float: left;" class="dropdown-item">
      <a ng-click="select(choice.id)">
        <angular-html-bind class="oppia-html-select-option protractor-test-html-select-option" html-data="choice.val">
        </angular-html-bind>
      </a>
    </li>
  </ul>
```

we get nice dropdown options that, while not identical to the production version, look nice:

![Screen Shot 2019-10-08 at 09 13 04](https://user-images.githubusercontent.com/19878639/66413238-dc869e00-e9ab-11e9-831d-c35503e8f43f.png)

I'll push a commit adding the `dropdown-item` class in a few hours. Using Angular UI Bootstrap with Bootstrap 4 might cause future issues though.

U8NWXD replied:
@DubeySandeep the lint tests timed out, but they pass when I run them locally. PTAL

ankita240796 replied:
I have restarted the tests once. 

I am not sure why the tests time out though. @kevinlee12 any idea of why this is happening?

DubeySandeep reviewed (commented):
@U8NWXD, Thanks for the PR, can you please add video/screenshot showcasing the changed UI?

ankita240796 commented:
Hi @U8NWXD, can you please look into this ASAP since this is blocking the Sept release? Thanks!

U8NWXD committed:
Replace Added CSS With dropdown-item Class

Boostrap 4 uses code like this for dropdown menus:

```html
<div class="dropdown-menu" aria-labelledby="dropdownMenuButton">
    <a class="dropdown-item" href="#">Action</a>
    <a class="dropdown-item" href="#">Another action</a>
    <a class="dropdown-item" href="#">Something else here</a>
</div>
```

With the upgrade from Bootstrap 3.3.4 to 4.3.1 in ba55ea9, the syntax
for dropdown menus changed to include the `dropdown-item` class for menu
items. The `html-select.directive.html` file was not updated, so this
commit adds the `dropdown-item` class to it.

Fixes #7681. Address review comments from @DubeySandeep.

U8NWXD committed:
Merge branch 'develop' of https://github.com/oppia/oppia into bugfix-html-select-options-style

ankita240796 commented:
Hi @DubeySandeep, can you PTAL. We need to get this in quickly for the release. Thanks!
